File: src/lib/algorithms/ordered.ts
// Ordered dithering using Bayer matrices
export function orderedDithering(
    grayscale: Uint8ClampedArray,
    width: number,
    height: number,
    dotSize: number = 1
  ): ImageData {
    // 4x4 Bayer matrix
    const bayerMatrix = [
      [ 0, 8, 2, 10],
      [12, 4, 14, 6],
      [ 3, 11, 1, 9],
      [15, 7, 13, 5]
    ];
    
    // Scale the Bayer matrix values from 0-15 to 0-255 range
    // For a 4x4 matrix, values are 0-15 (16 levels).
    // Scaled value = floor(original_value / 16 * 255)
    const bayerMatrixScaled = bayerMatrix.map(row => 
      row.map(val => Math.floor(val / 16 * 255))
    );
    
    // Create output ImageData
    const outputData = new ImageData(width, height);
    const output = outputData.data;
    
    // Apply dithering
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const idx = y * width + x;
        const pixelValue = grayscale[idx];
        
        // Get the Bayer threshold adjustment for this position
        const bayerX = Math.floor(x / dotSize) % 4;
        const bayerY = Math.floor(y / dotSize) % 4;
        const bayerValue = bayerMatrixScaled[bayerY][bayerX]; // Use the new scaled matrix
        
        // Apply threshold: compare pixel value with scaled Bayer matrix value
        const result = pixelValue > bayerValue ? 255 : 0;
        
        // Set all RGB channels to the result (black or white)
        output[idx * 4] = result;     // R
        output[idx * 4 + 1] = result; // G
        output[idx * 4 + 2] = result; // B
        output[idx * 4 + 3] = 255;    // A (fully opaque)
      }
    }
    
    return outputData;
  }

File: rust_wasm/src/ordered.rs
// Ordered dithering using Bayer matrices
pub fn ordered_dither(
    input_ptr: *const u8,
    width: usize,
    height: usize,
    dot_size: usize,
    output_ptr: *mut u8,
) {
    // 4x4 Bayer matrix
    const BAYER_MATRIX: [[u8; 4]; 4] = [
        [0, 8, 2, 10],
        [12, 4, 14, 6],
        [3, 11, 1, 9],
        [15, 7, 13, 5],
    ];
    
    // Convert the input and output pointers to slices
    let input = unsafe { std::slice::from_raw_parts(input_ptr, width * height) };
    let output = unsafe { std::slice::from_raw_parts_mut(output_ptr, width * height) };
    
    // Scale the Bayer matrix values to threshold range
    let mut bayer_scaled = [[0i16; 4]; 4];
    for y in 0..4 {
        for x in 0..4 {
            bayer_scaled[y][x] = ((BAYER_MATRIX[y][x] as f32 / 16.0) * 256.0 - 128.0) as i16;
        }
    }
    
    // Apply dithering
    for y in 0..height {
        for x in 0..width {
            let idx = y * width + x;
            let pixel_value = input[idx];
            
            // Get the Bayer threshold adjustment for this position
            let bayer_x = (x / dot_size) % 4;
            let bayer_y = (y / dot_size) % 4;
            let bayer_value = bayer_scaled[bayer_y][bayer_x];
            
            // Apply threshold with Bayer adjustment
            let result = if (pixel_value as i16 + bayer_value) > 128 { 255 } else { 0 };
            
            // Set the output pixel
            output[idx] = result;
        }
    }
} 